OsmAnd/src/net/osmand/plus/altimeter/AltimeterPlugin.java

```diff
@@ -0,0 +1,254 @@
+		return null;
+	}
+
+	@Override
+	public void registerLayers(MapActivity activity) {
+		registerWidget(activity);
+	}
+
+	private void registerWidget(MapActivity activity) {
+		MapInfoLayer mapInfoLayer = activity.getMapLayers().getMapInfoLayer();
+		if (mapInfoLayer != null ) {
+			altimeterControl = createAltimeterControl(activity);
+			mapInfoLayer.registerSideWidget(altimeterControl,
+								R.drawable.widget_altitude,  R.string.map_widget_altimeter, "altimeter", false, 21);
+			mapInfoLayer.recreateControls();
+			updateText("0");
+		}
+	}
+
+	private void updateText(String t) {
+		if(altimeterControl != null){
+			altimeterControl.setText(t,"");
+		}
+	}
+
+	private void showDialog(final MapActivity activity) {
+    Builder bld = new AlertDialog.Builder(activity);
+    LayoutInflater inflater = activity.getLayoutInflater();
+    View dialogView = inflater.inflate(R.layout.altimeter_dialog, null);
+
+    bld.setView(dialogView);
+    bld.setMessage(app.getString(R.string.map_widget_altimeter));
+    bld.setPositiveButton(R.string.shared_string_apply, new DialogInterface.OnClickListener() {
+        public void onClick(DialogInterface dialog, int id) {
+          qnh = qnhDialog;
+        }
+      });
+    bld.setNegativeButton(R.string.shared_string_cancel, new DialogInterface.OnClickListener() {
+        public void onClick(DialogInterface dialog, int id) {
+        }
+      });
+
+    edittextQNH = (EditText)dialogView.findViewById(R.id.edittext_qnh);
+    edittextAltitude = (EditText)dialogView.findViewById(R.id.edittext_altitude);
+
+    edittextQNH.addTextChangedListener(new TextWatcher(){
+          @Override
+          public void afterTextChanged(Editable s){}
+          @Override
+          public void beforeTextChanged(CharSequence s, int start, int count, int after){};
+          @Override
+          public void onTextChanged(CharSequence s, int start, int before, int count){
+            if(!promoteChange)return;
+            double altitude = 0;
+            double tmpQNH = QNH.standardAtmospherePressure;
+            try{
+              tmpQNH = Double.parseDouble(s.toString());
+            }catch(NumberFormatException e){
+              return;
+            }
+            qnhDialog = tmpQNH;
+            altitude = QNH.altitude(qnhDialog,pressure);
+            promoteChange = false;
+            edittextAltitude.setText(formatAltitude(altitude));
+            promoteChange = true;
+          }
+        });
+
+    edittextAltitude.addTextChangedListener(new TextWatcher(){
+          @Override
+          public void afterTextChanged(Editable s){}
+          @Override
+          public void beforeTextChanged(CharSequence s, int start, int count, int after){};
+          @Override
+          public void onTextChanged(CharSequence s, int start, int before, int count){
+            if(!promoteChange)return;
+            try{
+              qnhDialog = QNH.qnh(Double.parseDouble(s.toString()),pressure);
```

Comment:
Hi,
I am not sure that I understand your concern. I can simplify the first one (the one for edittextQNH.addTextChangedListener) to something like this (remove the variables altitude and tmpQNH):
```java
public void onTextChanged(CharSequence s, int start, int before, int count){
            if(!promoteChange)return;
            try{
              qnhDialog = Double.parseDouble(s.toString());
            }catch(NumberFormatException e){
              return;
            }
            promoteChange = false;
            edittextAltitude.setText(formatAltitude(QNH.altitude(qnhDialog,pressure)));
            promoteChange = true;
          }
```
I can add some comments. Or did you mean to expand the line:
```java
qnhDialog = QNH.qnh(Double.parseDouble(s.toString()),pressure) 
```
to something like:
```java
altitude = Double.parseDouble(s.toString());
qnhDialog = QNH.qnh(altitude,pressure);
```

?

The logic in both cases is simple - when the text in qnh or altitude changes, the other value is calculated and displayed. The effect is that as you type in one field the second is instantly changing to the appropriate value.

P.

Reply:
Ok, it's clearer with your explanations, thanks !

So for me here 2 things happen:
- convert qnh to altitude  (in QNH.java class) and convert string to double (and double to string)
- display the Strings in the EditText

But it should probably not be all at the same level.


It could probably be called using the same watcher, but specifying only the 2 variables: the text field to edit and the conversion direction.

```java
// currently, line 139 and 162
edittextQNH.addTextChangedListener(buildWatcherFor(edittextAltitude, s -> qnhToAltitude(s)));

edittextAltitude.addTextChangedListener(buildWatcherFor(edittextQNH, s -> altitudeToQNH(s)));
````

with a "generic" buildWatcherFor:
```java
    private TextWatcher buildWatcherFor(final EditText editTextToUpdate, final Function<CharSequence, String> conversionMethodToCall) {
        return new TextWatcher() {

            @Override
            public void afterTextChanged(Editable s) {
            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (!promoteChange) return;

                promoteChange = false;
                editTextToUpdate.setText(conversionMethodToCall.apply(s));
                promoteChange = true;
            }
        };
    }
```

I am not sure that the version of Java allows to pass a lambda as a parameter, but in this case a boolean can allow to format from qnh to altitude or vice-versa.

And then somewhere else (in this class ? or in QNH class?) you can implement the full method doing the qnhToAltitude and the altitudeToQnh.
(String to double parsing, the computation with scientific formula, and the double to String formatting)
But you will segregate the "business" about conversion, and the "display".

Reply:
Hi, thanks for explanation. I understand you concern better now :)  Since the java version is forced to 1.7 I refactored the code using an abstract class. I also commited changes for your other remarks. Please, let me know if it is OK now.

Thanks

Pavel